src/orion/algo/asha.py

```diff
@@ -0,0 +1,297 @@
+            point = list(self.space.sample(1, seed=tuple(self.rng.randint(0, 1000000, size=3)))[0])
+            if self.get_id(point) not in self.trial_info:
+                break
+
+        if self.get_id(point) in self.trial_info:
+            raise RuntimeError(
+                'ASHA keeps sampling already existing points. This should not happen, '
+                'please report this error to https://github.com/Epistimio/orion/issues')
+
+        sizes = numpy.array([len(b.rungs) for b in self.brackets])
+        probs = numpy.e**(sizes - sizes.max())
+        normalized = probs / probs.sum()
+        idx = self.rng.choice(len(self.brackets), p=normalized)
+
+        point[self.fidelity_index] = self.brackets[idx].rungs[0][0]
+
+        logger.debug('Sampling for bracket %s %s', idx, self.brackets[idx])
+
+        return [tuple(point)]
+
+    def get_id(self, point):
+        """Compute a unique hash for a point based on params, but not fidelity level."""
+        _point = list(point)
+        non_fidelity_dims = _point[0:self.fidelity_index]
+        non_fidelity_dims.extend(_point[self.fidelity_index + 1:])
+
+        return hashlib.md5(str(non_fidelity_dims).encode('utf-8')).hexdigest()
+
+    def observe(self, points, results):
+        """Observe evaluation `results` corresponding to list of `points` in
+        space.
+
+        A simple random sampler though does not take anything into account.
+        """
+        for point, result in zip(points, results):
+
+            _id = self.get_id(point)
+            bracket = self.trial_info.get(_id)
+
+            if not bracket:
+                fidelity = point[self.fidelity_index]
+                brackets = [bracket for bracket in self.brackets
+                            if bracket.rungs[0][0] == fidelity]
+                if not brackets:
+                    raise ValueError(
+                        "No bracket found for point {0} with fidelity {1}".format(_id, fidelity))
+                bracket = brackets[0]
+
+            try:
+                bracket.register(point, result['objective'])
+            except IndexError:
+                logger.warning('Point registered to wrong bracket. This is likely due '
+                               'to a corrupted database, where trials of different fidelity '
+                               'have a wrong timestamps.')
+                continue
+
+            if _id not in self.trial_info:
+                self.trial_info[_id] = bracket
+
+    @property
+    def is_done(self):
+        """Return True, if all brackets reached their maximum resources."""
+        return all(bracket.is_done for bracket in self.brackets)
+
+    @property
+    def fidelity_index(self):
+        """Compute the index of the point when fidelity is."""
+        def _is_fidelity(dim):
+            return (isinstance(dim, Fidelity) or
+                    (hasattr(dim, 'original_dimension') and
+                     isinstance(dim.original_dimension, Fidelity)))
```

Comment:
Just to make sure, this is for handling `TranformedSpace` right?

Reply:
Yep